src/cargo/util/config.rs

```diff
@@ -32,22 +33,125 @@ use util::Filesystem;
 
 use self::ConfigValue as CV;
 
+#[derive(Clone, Debug)]
+pub struct CargoDirs {
+    pub current_dir: PathBuf,
+    pub cache_dir: Filesystem,
+    pub config_dir: Filesystem,
+    pub data_dir: PathBuf,
+    pub bin_dir: PathBuf,
+}
+
+impl CargoDirs {
+    /// Computes the paths to directories used by cargo to retrieve and store
+    /// cache, config, ... files.
+    // This is written in the most straight-forward way possible, because it is
+    // hard as-is to understand all the different options, without trying to
+    // save lines of code.
+    pub fn new() -> CargoResult<CargoDirs> {
+        let current_dir = env::current_dir().chain_err(|| {
+            "couldn't get the current directory of the process"
+        })?;
+
+        let mut cache_dir  = PathBuf::default();
+        let mut config_dir = PathBuf::default();
+        let mut data_dir   = PathBuf::default();
+        let mut bin_dir    = PathBuf::default();
+
+        // 1. CARGO_HOME set
+        let cargo_home_env = env::var_os("CARGO_HOME").map(|home| current_dir.join(home));
+        if let Some(cargo_home) = cargo_home_env.clone() {
+            cache_dir  = cargo_home.clone();
+            config_dir = cargo_home.clone();
+            data_dir   = cargo_home.clone();
+            bin_dir    = cargo_home.join("bin");
+        }
+        // 2. CARGO_CACHE_DIR, CARGO_CONFIG_DIR, CARGO_BIN_DIR, ... set
+        let cargo_cache_env  = env::var_os("CARGO_CACHE_DIR").map(|home| current_dir.join(home));
+        let cargo_config_env = env::var_os("CARGO_CONFIG_DIR").map(|home| current_dir.join(home));
+        let cargo_data_env   = env::var_os("CARGO_DATA_DIR").map(|home| current_dir.join(home));
+        let cargo_bin_env    = env::var_os("CARGO_BIN_DIR").map(|home| current_dir.join(home));
+        if let Some(cargo_cache) = cargo_cache_env.clone() {
+            cache_dir = cargo_cache.clone();
+        }
+        if let Some(cargo_config) = cargo_config_env.clone() {
+            config_dir = cargo_config.clone();
+        }
+        if let Some(cargo_data) = cargo_data_env.clone() {
+            data_dir = cargo_data.clone();
+        }
+        if let Some(cargo_bin) = cargo_bin_env.clone() {
+            bin_dir = cargo_bin.clone();
+        }
+
+        // no env vars are set ...
+        if cargo_home_env.is_none()
+              && cargo_cache_env.is_none()
+              && cargo_config_env.is_none()
+              && cargo_data_env.is_none()
+              && cargo_bin_env.is_none() {
```

Comment:
This doesn't seem to handle the case where some but not all of the environment variables are set.